### PR TITLE
Improve Matrix Chain Multiplication

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
Can you have a look, @vil02 .

**Comment by TruongNhanNguyen** [comment]
Could you review this, @vil02.

File: src/dynamic_programming/matrix_chain_multiply.rs [hunk 1]
```diff
@@ -1,76 +1,74 @@
...
+///
+/// * `dimensions`: A vector where each element represents the dimensions of consecutive matrices
+///   in the chain. For example, [1, 2, 3, 4] represents matrices of dimensions (1x2), (2x3), and (3x4).
+///
+/// # Returns
+///
+/// The minimum number of scalar multiplications needed to compute the product of the matrices
+/// in the optimal order.
+pub fn matrix_chain_multiply(dimensions: Vec<usize>) -> usize {
+    if dimensions.len() <= 2 {
```
Comment (vil02) [inline src/dynamic_programming/matrix_chain_multiply.rs:25]:
```suggestion
    if dimensions.len() <= 1 {
```
The _single matrix_ input can be traded by the _general case_.

Furthermore I would suggest to _error_ for the  inputs of length 0 and 1. What does the input of a length 1 represent here? A _half matrix_?

File: src/dynamic_programming/matrix_chain_multiply.rs [hunk 2]
```diff
@@ -1,76 +1,74 @@
...
-        assert_eq!(matrix_chain_multiply(vec![1, 2, 3, 4, 3]), 30);
-        assert_eq!(matrix_chain_multiply(vec![1, 2, 3, 4, 3]), 30);
-        assert_eq!(matrix_chain_multiply(vec![4, 10, 3, 12, 20, 7]), 1344);
+    macro_rules! test_cases {
+        ($($name:ident: $test_case:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (input, expected) = $test_case;
+                    assert_eq!(matrix_chain_multiply(input), expected);
```
Comment (vil02) [inline src/dynamic_programming/matrix_chain_multiply.rs:58]: If you would like to _generate_ more test cases, you can do something like that:

```suggestion
                    assert_eq!(matrix_chain_multiply(input.clone()), expected);
                    assert_eq!(matrix_chain_multiply(input.into_iter().rev().collect()), expected);
```

File: src/dynamic_programming/matrix_chain_multiply.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/matrix_chain_multiply.rs:25]: Adding a custom error type is a good choice. I have added this error type to handle cases of inputs of length 0 and 1 length.
```rs
/// Custom error type for matrix chain multiplication.
#[derive(Debug, PartialEq)]
pub enum MatrixChainMultiplicationError {
    EmptyDimensions,
    InsufficientDimensions,
}
```

File: src/dynamic_programming/matrix_chain_multiply.rs (hunk 2 above)
Comment (TruongNhanNguyen) [reply src/dynamic_programming/matrix_chain_multiply.rs:58]: I will try it.

Comment (TruongNhanNguyen) [reply src/dynamic_programming/matrix_chain_multiply.rs:58]: All the tests passed.

File: src/dynamic_programming/matrix_chain_multiply.rs [hunk 3]
```diff
@@ -1,76 +1,92 @@
...
     }
 
-    #[test]
-    fn zero() {
-        assert_eq!(matrix_chain_multiply(vec![]), 0);
-        assert_eq!(matrix_chain_multiply(vec![10]), 0);
-        assert_eq!(matrix_chain_multiply(vec![10, 20]), 0);
+    test_cases! {
+        basic_chain_of_matrices: (vec![1, 2, 3, 4], Ok(18)),
+        matrices_in_reverse_order: (vec![4, 3, 2, 1], Ok(18)),
```
Comment (vil02) [inline src/dynamic_programming/matrix_chain_multiply.rs:84]:
```suggestion
```

Having 5860464a125678acec78dd4123a2cc8a70feaa52 makes this one redundant.

**Review by vil02 — APPROVED** [review]

**Pull request merged**